lib/postgrex_wal/gen_stage/pg_source.ex

```diff
@@ -0,0 +1,181 @@
+defmodule PostgrexWal.GenStage.PgSource do
+  @moduledoc ~S"""
+  ## Sample:
+  opts = [
+    source_name: :my_pg_source,
+    publication_name: "mypub1",
+    slot_name: "myslot1"
+    conn_opts: [host: "localhost", database: "r704_development", username: "jswk"]
+  ]
+  """
+  use Postgrex.ReplicationConnection, restart: :permanent, shutdown: 10_000
+  use TypedStruct
+  require Logger
+
+  typedstruct enforce: true do
+    field :publication_name, String.t()
+    field :slot_name, String.t()
+    field :final_lsn, integer(), default: 0
+    field :step, atom(), default: :disconnected
+  end
+
+  @typep conn_opts() :: [{:host, String.t()}, {:database, String.t()}, {:username, String.t()}]
+  @typep opts() :: [
+           {:source_name, String.t()},
+           {:publication_name, String.t()},
+           {:slot_name, String.t()},
+           {:conn_opts, conn_opts()}
+         ]
+  @spec start_link(opts()) :: {:ok, pid()} | {:error, Postgrex.Error.t() | term()}
+  def start_link(opts) do
+    # Automatically reconnect if we lose connection.
+    extra_opts = [
+      auto_reconnect: true,
+      name: opts[:source_name]
+    ]
+
+    Postgrex.ReplicationConnection.start_link(
+      __MODULE__,
+      struct(__MODULE__, Keyword.take(opts, [:publication_name, :slot_name])),
+      extra_opts ++ opts[:conn_opts]
+    )
+  end
+
+  # api functions()
+
+  @spec stop(pid) :: :ok
+  def stop(pid) do
+    GenServer.stop(pid)
+  end
+
+  @type lsn() :: integer() | String.t()
+  @type server() :: atom() | pid()
+  @spec async_ack(server(), lsn()) :: {:ack, lsn()}
+  def async_ack(server, lsn) when is_integer(lsn) do
+    send(server, {:ack, lsn})
+  end
+
+  def async_ack(server, lsn) when is_binary(lsn) do
+    {:ok, lsn} = Postgrex.ReplicationConnection.decode_lsn(lsn)
+    async_ack(server, lsn)
+  end
+
+  # callbacks()
+
+  @impl true
+  def init(state) do
+    {:ok, state}
+  end
+
+  @doc """
+  Replication slots provide an automated way to ensure that the primary does not remove WAL segments until
+  they have been received by all standbys, and that the primary does not remove rows which could cause a recovery
+  conflict even when the standby is disconnected.
+  """
+  @impl true
+  def handle_connect(state) do
+    query =
+      "START_REPLICATION SLOT #{state.slot_name} LOGICAL 0/0 (proto_version '2', publication_names '#{state.publication_name}')"
+
+    {:stream, query, [], struct(state, step: :streaming)}
```

Comment:
更新 map 或 struct 不用 `struct/2`
```suggestion
    {:stream, query, [], %{state | step: :streaming}}
```

Reply:
Done!